{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "Computing...\n",
      "CPU times: user 1min 20s, sys: 745 ms, total: 1min 21s\n",
      "Wall time: 1min 23s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "%load_ext autoreload\n",
    "\n",
    "from analysis_utils import *\n",
    "from joblib import Parallel, delayed\n",
    "import multiprocessing  \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "#Growth rate = 3.0\n",
    "#Computes the translation latencies from our model with simulations containing\n",
    "#1 to 14 cognate ternary complexes in the voxel (14 to 42 not included because\n",
    "#the probability of voxels with that number of cognates is ~0)\n",
    "path = \"/Users/Akshay/Documents/tRNAShuffle/data/\"\n",
    "ensmbl_latency_dict = dict()\n",
    "rxndiff=dict()\n",
    "scalingFactor = 200\n",
    "NR_scaledrates = {'k1r':718*200/scalingFactor,'k2f':1475*200/scalingFactor,'k2r_nr':1120*200/scalingFactor,'k3_nr':6*200/scalingFactor,'k4':209*200/scalingFactor}\n",
    "\n",
    "\n",
    "data = \"191003_2210/\"\n",
    "for i in range(1,7):\n",
    "    sim_i = computeTransportRxnTimes(path+data,100*(i-1),100*(i-1)+100,cogtRNANum=i, ribosomeNum = 7,scaling=scalingFactor,NR_scaling = NR_scaledrates)\n",
    "    ensmbl_latency_dict[i] = (CellLatencies(sim_i))\n",
    "\n",
    "data = \"191105_0908/\"\n",
    "for i in range(7,14):\n",
    "    sim_i = computeTransportRxnTimes(path+data, 100*(i-1-6),100*(i-1-6)+100,cogtRNANum=i, ribosomeNum = 7,scaling=scalingFactor,NR_scaling = NR_scaledrates)\n",
    "    ensmbl_latency_dict[i] = (CellLatencies(sim_i))\n",
    "    \n",
    "#Experimentally obtained codon usage frequencies and tRNA abundances at growth rate = 2.5 dbl/hr\n",
    "gr25_Codon = np.array([2.36, 1.26, 45.55, 34.17, 16.97, 57.86, 19.27, 33.74, 14.98, 22.31, 43.18, 7.67, 24.11, 24.87, 39.49, 11.81, 0.03, 0.63, 2.19, 9.31, 17.22, 55.01, 5.61, 29.21, 21.67, 0.52, 15.79, 43.86, 4.17, 2.61, 20.64, 26.7, 7.03, 0.19, 2.76, 3.81, 6.72, 16.52, 4.27, 2.73, 7.92, 23.25, 2.51, 1.98, 16.33, 11.68, 0.62, 0.67, 43.82, 20.59, 27.28, 7.01, 6.78, 14.21, 60.75, 0.82, 3.86, 4.09, 28.82, 5.18, 4.38, 1.09])\n",
    "gr25_pCodon = gr25_Codon/np.sum(gr25_Codon)\n",
    "gr25_ptRNA = [0.0602409638554217, 0.010542168674698798, 0.07379518072289158, 0.00602409638554217, 0.010542168674698798, 0.00602409638554217, 0.021084337349397596, 0.043674698795180725, 0.021084337349397596, 0.01204819277108434, 0.01807228915662651, 0.08433734939759038, 0.03162650602409639, 0.07228915662650603, 0.01204819277108434, 0.07078313253012049, 0.06325301204819278, 0.01656626506024097, 0.009036144578313254, 0.027108433734939763, 0.010542168674698798, 0.03012048192771085, 0.013554216867469882, 0.015060240963855425, 0.0075301204819277125, 0.010542168674698798, 0.0075301204819277125, 0.003012048192771085, 0.021084337349397596, 0.004518072289156627, 0.01656626506024097, 0.01204819277108434, 0.0015060240963855425, 0.009036144578313254, 0.01656626506024097, 0.019578313253012052, 0.015060240963855425, 0.027108434, 0.05873493975903616, 0.019578313]\n",
    "gr25_ptRNA = gr25_ptRNA/np.sum(gr25_ptRNA)\n",
    "\n",
    "n_cores = 16\n",
    "\n",
    "##Compute the codon distribution and elongation latency of each gene weighted by transcript number\n",
    "codon_elongt = compute_codon_elongt(gr25_ptRNA, len(gr25_pCodon), ensmbl_latency_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/akshay/opt/anaconda3/lib/python3.8/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elongt wildtype 172.03439339681063\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/akshay/opt/anaconda3/lib/python3.8/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elongt uniform 177.98557489079514\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/akshay/opt/anaconda3/lib/python3.8/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elongt codon-weighted correlated 162.55149505038742\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/akshay/opt/anaconda3/lib/python3.8/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elongt codon-weighted anticorrelated 201.5280117630773\n"
     ]
    }
   ],
   "source": [
    "#GFP gene optimization\n",
    "\n",
    "tRNA_distributions = {}\n",
    "tRNA_tags = ['Ala1B', 'Ala2', 'Arg2', 'Arg3', 'Arg4', 'Arg5', 'Asn', 'Asp1', 'Cys', 'Gln1', 'Gln2', 'Glu2', 'Gly2', 'Gly3', 'His', 'Ile1', 'Leu1', 'Leu2', 'Leu3', 'Leu4', 'Leu5', 'Lys', 'Met m', 'Phe', 'Pro1', 'Pro2', 'Pro3', 'Sec', 'Ser1', 'Ser2', 'Ser3', 'Ser5', 'Thr1', 'Thr2', 'Thr3', 'Thr4', 'Trp', 'Tyr1+Tyr2', 'Val1', 'Val2A+2B']\n",
    "codon_tags = ['GGG', 'GGA', 'GGU', 'GGC', 'GAG', 'GAA', 'GAU', 'GAC', 'GUG', 'GUA', 'GUU', 'GUC', 'GCG', 'GCA', 'GCU', 'GCC', 'AGG', 'AGA', 'AGU', 'AGC', 'AAG', 'AAA', 'AAU', 'AAC', 'AUG', 'AUA', 'AUU', 'AUC', 'ACG', 'ACA', 'ACU', 'ACC', 'UGG', 'UGA', 'UGU', 'UGC', 'UAU', 'UAC', 'UUG', 'UUA', 'UUU', 'UUC', 'UCG', 'UCA', 'UCU', 'UCC', 'CGG', 'CGA', 'CGU', 'CGC', 'CAG', 'CAA', 'CAU', 'CAC', 'CUG', 'CUA', 'CUU', 'CUC', 'CCG', 'CCA', 'CCU', 'CCC']\n",
    "\n",
    "\n",
    "from analysis_utils import *\n",
    "gene_map, pCodon_GFP = singlegene_to_genemap('/Users/akshay/Documents/tRNAShuffle/data/tables/sfGFP_RED20.csv')\n",
    "#wildtype\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "\n",
    "codon_elongt = compute_codon_elongt(ptRNA_red20, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_wt = compute_gene_elongt(gene_map['sfGFP_RED20'],codon_elongt)\n",
    "print('elongt wildtype', avg_elongt_wt)\n",
    "tRNA_distributions['GFP_wt'] = [avg_elongt_wt,dict(zip(tRNA_tags,ptRNA_red20)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "#Uniform\n",
    "ptRNA_red20_uniform = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20_uniform = ptRNA_red20_uniform[6]/sum(ptRNA_red20_uniform[6])\n",
    "ptRNA_red20_uniform[ptRNA_red20_uniform>0] = 1/20\n",
    "\n",
    "codon_elongt = compute_codon_elongt(ptRNA_red20_uniform, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_uniform = compute_gene_elongt(gene_map['sfGFP_RED20'],codon_elongt)\n",
    "print('elongt uniform', avg_elongt_uniform)\n",
    "tRNA_distributions['GFP_uniform'] = [avg_elongt_uniform,dict(zip(tRNA_tags,ptRNA_red20_uniform)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "\n",
    "#GFP Codon-weighted correlated\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "ptRNA_dict, pcodon_dict, codon_dict, codonLabels, pCodon, tRNA_dict = cognateDistrib(ptRNA_red20,pCodon_GFP,extra2=True)\n",
    "tRNA_sum_dict = tRNA_dict.copy()\n",
    "tRNA_weightedbycodon = list()\n",
    "for keys in tRNA_dict.keys():\n",
    "    tRNA_sum_dict[keys] = (sum([pcodon_dict[tRNA_key] for tRNA_key in tRNA_dict[keys]]))\n",
    "tRNA_weightedbycodon = list(tRNA_sum_dict.values())/sum(tRNA_sum_dict.values())\n",
    "tRNA_sorted_indices = np.flip(tRNA_weightedbycodon.argsort())\n",
    "tRNA_sorted = np.array(tRNA_tags)[tRNA_sorted_indices]\n",
    "\n",
    "##Assign monotonic increase to GFP codon-sorted tRNA\n",
    "ptRNA_sorted = tRNA_weightedbycodon[tRNA_sorted_indices]\n",
    "sortedtRNAdict = dict(zip(tRNA_sorted, ptRNA_sorted))\n",
    "resortedtRNA = [sortedtRNAdict[tRNA] for tRNA in tRNA_tags]\n",
    "\n",
    "codon_elongt = compute_codon_elongt(resortedtRNA, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_codonweightedcorrelated = compute_gene_elongt(gene_map['sfGFP_RED20'],codon_elongt)\n",
    "print('elongt codon-weighted correlated', avg_elongt_codonweightedcorrelated)\n",
    "\n",
    "tRNA_distributions['GFP_codonweightedcorrelated'] = [avg_elongt_codonweightedcorrelated,dict(zip(tRNA_tags,resortedtRNA)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "\n",
    "#GFP Codon-weighted anticorrelated\n",
    "tRNA_tags = ['Ala1B', 'Ala2', 'Arg2', 'Arg3', 'Arg4', 'Arg5', 'Asn', 'Asp1', 'Cys', 'Gln1', 'Gln2', 'Glu2', 'Gly2', 'Gly3', 'His', 'Ile1', 'Leu1', 'Leu2', 'Leu3', 'Leu4', 'Leu5', 'Lys', 'Met m', 'Phe', 'Pro1', 'Pro2', 'Pro3', 'Sec', 'Ser1', 'Ser2', 'Ser3', 'Ser5', 'Thr1', 'Thr2', 'Thr3', 'Thr4', 'Trp', 'Tyr1+Tyr2', 'Val1', 'Val2A+2B']\n",
    "codon_tags = ['GGG', 'GGA', 'GGU', 'GGC', 'GAG', 'GAA', 'GAU', 'GAC', 'GUG', 'GUA', 'GUU', 'GUC', 'GCG', 'GCA', 'GCU', 'GCC', 'AGG', 'AGA', 'AGU', 'AGC', 'AAG', 'AAA', 'AAU', 'AAC', 'AUG', 'AUA', 'AUU', 'AUC', 'ACG', 'ACA', 'ACU', 'ACC', 'UGG', 'UGA', 'UGU', 'UGC', 'UAU', 'UAC', 'UUG', 'UUA', 'UUU', 'UUC', 'UCG', 'UCA', 'UCU', 'UCC', 'CGG', 'CGA', 'CGU', 'CGC', 'CAG', 'CAA', 'CAU', 'CAC', 'CUG', 'CUA', 'CUU', 'CUC', 'CCG', 'CCA', 'CCU', 'CCC']\n",
    "\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "ptRNA_dict, pcodon_dict, codon_dict, codonLabels, pCodon, tRNA_dict = cognateDistrib(ptRNA_red20,pCodon_GFP,extra2=True)\n",
    "tRNA_sum_dict = tRNA_dict.copy()\n",
    "tRNA_weightedbycodon = list()\n",
    "for keys in tRNA_dict.keys():\n",
    "    tRNA_sum_dict[keys] = (sum([pcodon_dict[tRNA_key] for tRNA_key in tRNA_dict[keys]]))\n",
    "tRNA_weightedbycodon = list(tRNA_sum_dict.values())/sum(tRNA_sum_dict.values())\n",
    "tRNA_sorted_indices = np.flip(tRNA_weightedbycodon.argsort())\n",
    "tRNA_sorted = np.array(tRNA_tags)[tRNA_sorted_indices]\n",
    "\n",
    "##Assign monotonic increase to GFP codon-sorted tRNA\n",
    "ptRNA_sorted = tRNA_weightedbycodon[tRNA_sorted_indices]\n",
    "ptRNA_sorted_flipped = np.append(np.flip([ptRNA for ptRNA in ptRNA_sorted if ptRNA != 0]),np.array([ptRNA for ptRNA in ptRNA_sorted if ptRNA == 0]))\n",
    "sortedtRNAdict = dict(zip(tRNA_sorted, ptRNA_sorted_flipped))\n",
    "resortedtRNA = [sortedtRNAdict[tRNA] for tRNA in tRNA_tags]\n",
    "\n",
    "codon_elongt = compute_codon_elongt(resortedtRNA, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_codonweightedanticorrelated = compute_gene_elongt(gene_map['sfGFP_RED20'],codon_elongt)\n",
    "print('elongt codon-weighted anticorrelated', avg_elongt_codonweightedanticorrelated)\n",
    "\n",
    "tRNA_distributions['GFP_codonweightedanticorrelated'] = [avg_elongt_codonweightedanticorrelated,dict(zip(tRNA_tags,resortedtRNA)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "\n",
    "########### Genetic algorithm fast and slow #############\n",
    "\n",
    "elongt_list_fast =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_fasttRNA_elongt.npy')\n",
    "ptRNA_list_fast=np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_fasttRNA_ptRNA.npy')\n",
    "elongt_fastest = np.min(elongt_list_fast)\n",
    "ptRNA_min = ptRNA_list_fast[np.argmin(elongt_list_fast)]\n",
    "tRNA_distributions['GFP_gafast'] = [elongt_fastest,dict(zip(tRNA_tags,ptRNA_min)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "elongt_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_slowtRNA_elongt.npy')\n",
    "ptRNA_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_slowtRNA_ptRNA.npy')\n",
    "elongt_slowest = np.max(elongt_list_slow)\n",
    "ptRNA_max = ptRNA_list_slow[np.argmax(elongt_list_slow)]\n",
    "tRNA_distributions['GFP_gaslow'] = [elongt_slowest,dict(zip(tRNA_tags,ptRNA_max)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "\n",
    "######### Export all to pickle ##############\n",
    "\n",
    "import pickle\n",
    "with open('tRNA_distributions-GFP.pickle', 'wb') as handle:\n",
    "    pickle.dump(tRNA_distributions, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[250.83730445466912,\n",
       " {'Ala1B': 0.00661085002001485,\n",
       "  'Ala2': 0.0,\n",
       "  'Arg2': 0.0019971128079857344,\n",
       "  'Arg3': 0.0,\n",
       "  'Arg4': 0.0,\n",
       "  'Arg5': 0.0,\n",
       "  'Asn': 0.0014292634719609955,\n",
       "  'Asp1': 0.003766698270230139,\n",
       "  'Cys': 0.0047040969672237235,\n",
       "  'Gln1': 0.0,\n",
       "  'Gln2': 0.002308752464957013,\n",
       "  'Glu2': 0.007053452230498484,\n",
       "  'Gly2': 0.002656256890993093,\n",
       "  'Gly3': 0.0,\n",
       "  'His': 0.005606656731590659,\n",
       "  'Ile1': 0.0031969596152049843,\n",
       "  'Leu1': 0.0,\n",
       "  'Leu2': 0.0,\n",
       "  'Leu3': 0.0027200038313679614,\n",
       "  'Leu4': 0.0,\n",
       "  'Leu5': 0.0,\n",
       "  'Lys': 0.0022994414824551866,\n",
       "  'Met m': 0.006007177101277787,\n",
       "  'Phe': 0.006364825205384883,\n",
       "  'Pro1': 0.6772020061619743,\n",
       "  'Pro2': 0.0,\n",
       "  'Pro3': 0.0,\n",
       "  'Sec': 0.0,\n",
       "  'Ser1': 0.001964793853604591,\n",
       "  'Ser2': 0.0,\n",
       "  'Ser3': 0.0,\n",
       "  'Ser5': 0.0,\n",
       "  'Thr1': 0.0027545724067898647,\n",
       "  'Thr2': 0.0,\n",
       "  'Thr3': 0.0,\n",
       "  'Thr4': 0.0,\n",
       "  'Trp': 0.2549425809848878,\n",
       "  'Tyr1+Tyr2': 0.003329881490910349,\n",
       "  'Val1': 0.0,\n",
       "  'Val2A+2B': 0.003084618010687356},\n",
       " {'GGG': 0.0,\n",
       "  'GGA': 0.09205020920502092,\n",
       "  'GGU': 0.0,\n",
       "  'GGC': 0.0,\n",
       "  'GAG': 0.0,\n",
       "  'GAA': 0.06694560669456066,\n",
       "  'GAU': 0.0,\n",
       "  'GAC': 0.07531380753138076,\n",
       "  'GUG': 0.0,\n",
       "  'GUA': 0.0,\n",
       "  'GUU': 0.0,\n",
       "  'GUC': 0.07531380753138076,\n",
       "  'GCG': 0.0,\n",
       "  'GCA': 0.03347280334728033,\n",
       "  'GCU': 0.0,\n",
       "  'GCC': 0.0,\n",
       "  'AGG': 0.0,\n",
       "  'AGA': 0.0,\n",
       "  'AGU': 0.0,\n",
       "  'AGC': 0.0,\n",
       "  'AAG': 0.0,\n",
       "  'AAA': 0.08368200836820083,\n",
       "  'AAU': 0.0,\n",
       "  'AAC': 0.05439330543933055,\n",
       "  'AUG': 0.02092050209205021,\n",
       "  'AUA': 0.0,\n",
       "  'AUU': 0.0,\n",
       "  'AUC': 0.04602510460251046,\n",
       "  'ACG': 0.0,\n",
       "  'ACA': 0.0,\n",
       "  'ACU': 0.0,\n",
       "  'ACC': 0.07531380753138076,\n",
       "  'UGG': 0.0041841004184100415,\n",
       "  'UGA': 0.0041841004184100415,\n",
       "  'UGU': 0.0,\n",
       "  'UGC': 0.008368200836820083,\n",
       "  'UAU': 0.0,\n",
       "  'UAC': 0.03765690376569038,\n",
       "  'UUG': 0.0,\n",
       "  'UUA': 0.0,\n",
       "  'UUU': 0.0,\n",
       "  'UUC': 0.0502092050209205,\n",
       "  'UCG': 0.0,\n",
       "  'UCA': 0.04184100418410042,\n",
       "  'UCU': 0.0,\n",
       "  'UCC': 0.0,\n",
       "  'CGG': 0.0,\n",
       "  'CGA': 0.0,\n",
       "  'CGU': 0.03347280334728033,\n",
       "  'CGC': 0.0,\n",
       "  'CAG': 0.029288702928870293,\n",
       "  'CAA': 0.0,\n",
       "  'CAU': 0.0,\n",
       "  'CAC': 0.04184100418410042,\n",
       "  'CUG': 0.0,\n",
       "  'CUA': 0.08368200836820083,\n",
       "  'CUU': 0.0,\n",
       "  'CUC': 0.0,\n",
       "  'CCG': 0.04184100418410042,\n",
       "  'CCA': 0.0,\n",
       "  'CCU': 0.0,\n",
       "  'CCC': 0.0}]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_distributions['GFP_gaslow']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elongt wildtype 173.44537806058804\n",
      "elongt uniform 179.08324128594163\n",
      "elongt codon-weighted correlated 162.42524525183458\n",
      "elongt codon-weighted anticorrelated 196.00742981509876\n"
     ]
    }
   ],
   "source": [
    "#mRFP gene optimization\n",
    "\n",
    "tRNA_distributions = {}\n",
    "tRNA_tags = ['Ala1B', 'Ala2', 'Arg2', 'Arg3', 'Arg4', 'Arg5', 'Asn', 'Asp1', 'Cys', 'Gln1', 'Gln2', 'Glu2', 'Gly2', 'Gly3', 'His', 'Ile1', 'Leu1', 'Leu2', 'Leu3', 'Leu4', 'Leu5', 'Lys', 'Met m', 'Phe', 'Pro1', 'Pro2', 'Pro3', 'Sec', 'Ser1', 'Ser2', 'Ser3', 'Ser5', 'Thr1', 'Thr2', 'Thr3', 'Thr4', 'Trp', 'Tyr1+Tyr2', 'Val1', 'Val2A+2B']\n",
    "codon_tags = ['GGG', 'GGA', 'GGU', 'GGC', 'GAG', 'GAA', 'GAU', 'GAC', 'GUG', 'GUA', 'GUU', 'GUC', 'GCG', 'GCA', 'GCU', 'GCC', 'AGG', 'AGA', 'AGU', 'AGC', 'AAG', 'AAA', 'AAU', 'AAC', 'AUG', 'AUA', 'AUU', 'AUC', 'ACG', 'ACA', 'ACU', 'ACC', 'UGG', 'UGA', 'UGU', 'UGC', 'UAU', 'UAC', 'UUG', 'UUA', 'UUU', 'UUC', 'UCG', 'UCA', 'UCU', 'UCC', 'CGG', 'CGA', 'CGU', 'CGC', 'CAG', 'CAA', 'CAU', 'CAC', 'CUG', 'CUA', 'CUU', 'CUC', 'CCG', 'CCA', 'CCU', 'CCC']\n",
    "\n",
    "\n",
    "from analysis_utils import *\n",
    "gene_map, pCodon_mRFP = singlegene_to_genemap('/Users/akshay/Documents/tRNAShuffle/data/tables/mRFP_RED20.csv')\n",
    "#wildtype\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "\n",
    "codon_elongt = compute_codon_elongt(ptRNA_red20, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_wt = compute_gene_elongt(gene_map['mRFP_RED20'],codon_elongt)\n",
    "print('elongt wildtype', avg_elongt_wt)\n",
    "tRNA_distributions['mRFP_RED20'] = [avg_elongt_wt,dict(zip(tRNA_tags,ptRNA_red20)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "#Uniform\n",
    "ptRNA_red20_uniform = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20_uniform = ptRNA_red20_uniform[6]/sum(ptRNA_red20_uniform[6])\n",
    "ptRNA_red20_uniform[ptRNA_red20_uniform>0] = 1/20\n",
    "\n",
    "codon_elongt = compute_codon_elongt(ptRNA_red20_uniform, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_uniform = compute_gene_elongt(gene_map['mRFP_RED20'],codon_elongt)\n",
    "print('elongt uniform', avg_elongt_uniform)\n",
    "tRNA_distributions['mRFP_uniform'] = [avg_elongt_uniform,dict(zip(tRNA_tags,ptRNA_red20_uniform)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "\n",
    "#GFP Codon-weighted correlated\n",
    "\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "ptRNA_dict, pcodon_dict, codon_dict, codonLabels, pCodon, tRNA_dict = cognateDistrib(ptRNA_red20,pCodon_mRFP,extra2=True)\n",
    "tRNA_sum_dict = tRNA_dict.copy()\n",
    "tRNA_weightedbycodon = list()\n",
    "for keys in tRNA_dict.keys():\n",
    "    tRNA_sum_dict[keys] = (sum([pcodon_dict[tRNA_key] for tRNA_key in tRNA_dict[keys]]))\n",
    "tRNA_weightedbycodon = list(tRNA_sum_dict.values())/sum(tRNA_sum_dict.values())\n",
    "tRNA_sorted_indices = np.flip(tRNA_weightedbycodon.argsort())\n",
    "tRNA_sorted = np.array(tRNA_tags)[tRNA_sorted_indices]\n",
    "\n",
    "##Assign monotonic increase to GFP codon-sorted tRNA\n",
    "ptRNA_sorted = tRNA_weightedbycodon[tRNA_sorted_indices]\n",
    "sortedtRNAdict = dict(zip(tRNA_sorted, ptRNA_sorted))\n",
    "resortedtRNA = [sortedtRNAdict[tRNA] for tRNA in tRNA_tags]\n",
    "\n",
    "codon_elongt = compute_codon_elongt(resortedtRNA, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_codonweightedcorrelated = compute_gene_elongt(gene_map['mRFP_RED20'],codon_elongt)\n",
    "print('elongt codon-weighted correlated', avg_elongt_codonweightedcorrelated)\n",
    "\n",
    "tRNA_distributions['mRFP_codonweightedcorrelated'] = [avg_elongt_codonweightedcorrelated,dict(zip(tRNA_tags,resortedtRNA)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "\n",
    "#GFP Codon-weighted anticorrelated\n",
    "tRNA_tags = ['Ala1B', 'Ala2', 'Arg2', 'Arg3', 'Arg4', 'Arg5', 'Asn', 'Asp1', 'Cys', 'Gln1', 'Gln2', 'Glu2', 'Gly2', 'Gly3', 'His', 'Ile1', 'Leu1', 'Leu2', 'Leu3', 'Leu4', 'Leu5', 'Lys', 'Met m', 'Phe', 'Pro1', 'Pro2', 'Pro3', 'Sec', 'Ser1', 'Ser2', 'Ser3', 'Ser5', 'Thr1', 'Thr2', 'Thr3', 'Thr4', 'Trp', 'Tyr1+Tyr2', 'Val1', 'Val2A+2B']\n",
    "codon_tags = ['GGG', 'GGA', 'GGU', 'GGC', 'GAG', 'GAA', 'GAU', 'GAC', 'GUG', 'GUA', 'GUU', 'GUC', 'GCG', 'GCA', 'GCU', 'GCC', 'AGG', 'AGA', 'AGU', 'AGC', 'AAG', 'AAA', 'AAU', 'AAC', 'AUG', 'AUA', 'AUU', 'AUC', 'ACG', 'ACA', 'ACU', 'ACC', 'UGG', 'UGA', 'UGU', 'UGC', 'UAU', 'UAC', 'UUG', 'UUA', 'UUU', 'UUC', 'UCG', 'UCA', 'UCU', 'UCC', 'CGG', 'CGA', 'CGU', 'CGC', 'CAG', 'CAA', 'CAU', 'CAC', 'CUG', 'CUA', 'CUU', 'CUC', 'CCG', 'CCA', 'CCU', 'CCC']\n",
    "\n",
    "ptRNA_red20 = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20 = ptRNA_red20[6]/sum(ptRNA_red20[6])\n",
    "ptRNA_dict, pcodon_dict, codon_dict, codonLabels, pCodon, tRNA_dict = cognateDistrib(ptRNA_red20,pCodon_mRFP,extra2=True)\n",
    "tRNA_sum_dict = tRNA_dict.copy()\n",
    "tRNA_weightedbycodon = list()\n",
    "for keys in tRNA_dict.keys():\n",
    "    tRNA_sum_dict[keys] = (sum([pcodon_dict[tRNA_key] for tRNA_key in tRNA_dict[keys]]))\n",
    "tRNA_weightedbycodon = list(tRNA_sum_dict.values())/sum(tRNA_sum_dict.values())\n",
    "tRNA_sorted_indices = np.flip(tRNA_weightedbycodon.argsort())\n",
    "tRNA_sorted = np.array(tRNA_tags)[tRNA_sorted_indices]\n",
    "\n",
    "##Assign monotonic increase to GFP codon-sorted tRNA\n",
    "ptRNA_sorted = tRNA_weightedbycodon[tRNA_sorted_indices]\n",
    "ptRNA_sorted_flipped = np.append(np.flip([ptRNA for ptRNA in ptRNA_sorted if ptRNA != 0]),np.array([ptRNA for ptRNA in ptRNA_sorted if ptRNA == 0]))\n",
    "sortedtRNAdict = dict(zip(tRNA_sorted, ptRNA_sorted_flipped))\n",
    "resortedtRNA = [sortedtRNAdict[tRNA] for tRNA in tRNA_tags]\n",
    "\n",
    "codon_elongt = compute_codon_elongt(resortedtRNA, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_codonweightedanticorrelated = compute_gene_elongt(gene_map['mRFP_RED20'],codon_elongt)\n",
    "print('elongt codon-weighted anticorrelated', avg_elongt_codonweightedanticorrelated)\n",
    "\n",
    "tRNA_distributions['mRFP_codonweightedanticorrelated'] = [avg_elongt_codonweightedanticorrelated,dict(zip(tRNA_tags,resortedtRNA)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "\n",
    "########### Genetic algorithm fast and slow #############\n",
    "\n",
    "elongt_list_fast =np.load('./data/UniformRandomtRNA_minimalcode-mRFP_2ksample_fasttRNA_elongt.npy')\n",
    "ptRNA_list_fast=np.load('./data/UniformRandomtRNA_minimalcode-mRFP_2ksample_fasttRNA_ptRNA.npy')\n",
    "elongt_fastest = np.min(elongt_list_fast)\n",
    "ptRNA_min = ptRNA_list_fast[np.argmin(elongt_list_fast)]\n",
    "tRNA_distributions['mRFP_gafast'] = [elongt_fastest,dict(zip(tRNA_tags,ptRNA_min)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "elongt_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-mRFP_2ksample_slowtRNA_elongt.npy')\n",
    "ptRNA_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-mRFP_2ksample_slowtRNA_ptRNA.npy')\n",
    "elongt_slowest = np.max(elongt_list_slow)\n",
    "ptRNA_max = ptRNA_list_slow[np.argmax(elongt_list_slow)]\n",
    "tRNA_distributions['mRFP_gaslow'] = [elongt_slowest,dict(zip(tRNA_tags,ptRNA_max)),dict(zip(codon_tags,pCodon_mRFP))]\n",
    "\n",
    "\n",
    "######### Export all to pickle ##############\n",
    "\n",
    "import pickle\n",
    "with open('tRNA_distributions-mRFP.pickle', 'wb') as handle:\n",
    "    pickle.dump(tRNA_distributions, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Max elongt check w/ all tRNA zero but one tRNA = 1 probability, as control\n",
    "ptRNA_red20_zeros = pd.read_excel('./data/tables/tRNAValues_RED20.xlsx',header=None)\n",
    "ptRNA_red20_zeros = ptRNA_red20_zeros[6]/sum(ptRNA_red20_zeros[6])\n",
    "ptRNA_red20_zeros[ptRNA_red20_zeros>0] = 0\n",
    "ptRNA_red20_zeros[1] = 1\n",
    "\n",
    "codon_elongt = compute_codon_elongt(ptRNA_red20_zeros, len(pCodon_GFP), ensmbl_latency_dict)\n",
    "avg_elongt_max = compute_gene_elongt(gene_map['sfGFP_RED20'],codon_elongt)\n",
    "print('elongt max', avg_elongt_max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open(r\"tRNA_distributions-GFP.pickle\", \"rb\") as input_file:\n",
    "    e = pickle.load(input_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[250.83730445466912,\n",
       " {'Ala1B': 0.00661085002001485,\n",
       "  'Ala2': 0.0,\n",
       "  'Arg2': 0.0019971128079857344,\n",
       "  'Arg3': 0.0,\n",
       "  'Arg4': 0.0,\n",
       "  'Arg5': 0.0,\n",
       "  'Asn': 0.0014292634719609955,\n",
       "  'Asp1': 0.003766698270230139,\n",
       "  'Cys': 0.0047040969672237235,\n",
       "  'Gln1': 0.0,\n",
       "  'Gln2': 0.002308752464957013,\n",
       "  'Glu2': 0.007053452230498484,\n",
       "  'Gly2': 0.002656256890993093,\n",
       "  'Gly3': 0.0,\n",
       "  'His': 0.005606656731590659,\n",
       "  'Ile1': 0.0031969596152049843,\n",
       "  'Leu1': 0.0,\n",
       "  'Leu2': 0.0,\n",
       "  'Leu3': 0.0027200038313679614,\n",
       "  'Leu4': 0.0,\n",
       "  'Leu5': 0.0,\n",
       "  'Lys': 0.0022994414824551866,\n",
       "  'Met m': 0.006007177101277787,\n",
       "  'Phe': 0.006364825205384883,\n",
       "  'Pro1': 0.6772020061619743,\n",
       "  'Pro2': 0.0,\n",
       "  'Pro3': 0.0,\n",
       "  'Sec': 0.0,\n",
       "  'Ser1': 0.001964793853604591,\n",
       "  'Ser2': 0.0,\n",
       "  'Ser3': 0.0,\n",
       "  'Ser5': 0.0,\n",
       "  'Thr1': 0.0027545724067898647,\n",
       "  'Thr2': 0.0,\n",
       "  'Thr3': 0.0,\n",
       "  'Thr4': 0.0,\n",
       "  'Trp': 0.2549425809848878,\n",
       "  'Tyr1+Tyr2': 0.003329881490910349,\n",
       "  'Val1': 0.0,\n",
       "  'Val2A+2B': 0.003084618010687356},\n",
       " {'GGG': 0.0,\n",
       "  'GGA': 0.09205020920502092,\n",
       "  'GGU': 0.0,\n",
       "  'GGC': 0.0,\n",
       "  'GAG': 0.0,\n",
       "  'GAA': 0.06694560669456066,\n",
       "  'GAU': 0.0,\n",
       "  'GAC': 0.07531380753138076,\n",
       "  'GUG': 0.0,\n",
       "  'GUA': 0.0,\n",
       "  'GUU': 0.0,\n",
       "  'GUC': 0.07531380753138076,\n",
       "  'GCG': 0.0,\n",
       "  'GCA': 0.03347280334728033,\n",
       "  'GCU': 0.0,\n",
       "  'GCC': 0.0,\n",
       "  'AGG': 0.0,\n",
       "  'AGA': 0.0,\n",
       "  'AGU': 0.0,\n",
       "  'AGC': 0.0,\n",
       "  'AAG': 0.0,\n",
       "  'AAA': 0.08368200836820083,\n",
       "  'AAU': 0.0,\n",
       "  'AAC': 0.05439330543933055,\n",
       "  'AUG': 0.02092050209205021,\n",
       "  'AUA': 0.0,\n",
       "  'AUU': 0.0,\n",
       "  'AUC': 0.04602510460251046,\n",
       "  'ACG': 0.0,\n",
       "  'ACA': 0.0,\n",
       "  'ACU': 0.0,\n",
       "  'ACC': 0.07531380753138076,\n",
       "  'UGG': 0.0041841004184100415,\n",
       "  'UGA': 0.0041841004184100415,\n",
       "  'UGU': 0.0,\n",
       "  'UGC': 0.008368200836820083,\n",
       "  'UAU': 0.0,\n",
       "  'UAC': 0.03765690376569038,\n",
       "  'UUG': 0.0,\n",
       "  'UUA': 0.0,\n",
       "  'UUU': 0.0,\n",
       "  'UUC': 0.0502092050209205,\n",
       "  'UCG': 0.0,\n",
       "  'UCA': 0.04184100418410042,\n",
       "  'UCU': 0.0,\n",
       "  'UCC': 0.0,\n",
       "  'CGG': 0.0,\n",
       "  'CGA': 0.0,\n",
       "  'CGU': 0.03347280334728033,\n",
       "  'CGC': 0.0,\n",
       "  'CAG': 0.029288702928870293,\n",
       "  'CAA': 0.0,\n",
       "  'CAU': 0.0,\n",
       "  'CAC': 0.04184100418410042,\n",
       "  'CUG': 0.0,\n",
       "  'CUA': 0.08368200836820083,\n",
       "  'CUU': 0.0,\n",
       "  'CUC': 0.0,\n",
       "  'CCG': 0.04184100418410042,\n",
       "  'CCA': 0.0,\n",
       "  'CCU': 0.0,\n",
       "  'CCC': 0.0}]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e['GFP_gaslow']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GFP GA slow fast optimization\n",
    "\n",
    "tRNA_distributions = {}\n",
    "from analysis_utils import *\n",
    "gene_map, pCodon_GFP = singlegene_to_genemap('/Users/akshay/Documents/tRNAShuffle/data/tables/sfGFP_RED20.csv')\n",
    "\n",
    "tRNA_tags = ['Ala1B', 'Ala2', 'Arg2', 'Arg3', 'Arg4', 'Arg5', 'Asn', 'Asp1', 'Cys', 'Gln1', 'Gln2', 'Glu2', 'Gly2', 'Gly3', 'His', 'Ile1', 'Leu1', 'Leu2', 'Leu3', 'Leu4', 'Leu5', 'Lys', 'Met m', 'Phe', 'Pro1', 'Pro2', 'Pro3', 'Sec', 'Ser1', 'Ser2', 'Ser3', 'Ser5', 'Thr1', 'Thr2', 'Thr3', 'Thr4', 'Trp', 'Tyr1+Tyr2', 'Val1', 'Val2A+2B']\n",
    "codon_tags = ['GGG', 'GGA', 'GGU', 'GGC', 'GAG', 'GAA', 'GAU', 'GAC', 'GUG', 'GUA', 'GUU', 'GUC', 'GCG', 'GCA', 'GCU', 'GCC', 'AGG', 'AGA', 'AGU', 'AGC', 'AAG', 'AAA', 'AAU', 'AAC', 'AUG', 'AUA', 'AUU', 'AUC', 'ACG', 'ACA', 'ACU', 'ACC', 'UGG', 'UGA', 'UGU', 'UGC', 'UAU', 'UAC', 'UUG', 'UUA', 'UUU', 'UUC', 'UCG', 'UCA', 'UCU', 'UCC', 'CGG', 'CGA', 'CGU', 'CGC', 'CAG', 'CAA', 'CAU', 'CAC', 'CUG', 'CUA', 'CUU', 'CUC', 'CCG', 'CCA', 'CCU', 'CCC']\n",
    "\n",
    "import numpy as np\n",
    "elongt_list_fast =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_fasttRNA_elongt.npy')\n",
    "ptRNA_list_fast=np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_fasttRNA_ptRNA.npy')\n",
    "elongt_fastest = np.min(elongt_list_fast)\n",
    "ptRNA_min = ptRNA_list_fast[np.argmin(elongt_list_fast)]\n",
    "tRNA_distributions['GFP_gafast'] = [elongt_fastest,dict(zip(tRNA_tags,ptRNA_min)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "elongt_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_slowtRNA_elongt.npy')\n",
    "ptRNA_list_slow =np.load('./data/UniformRandomtRNA_minimalcode-sfGFP_2ksample_slowtRNA_ptRNA.npy')\n",
    "elongt_slowest = np.max(elongt_list_slow)\n",
    "ptRNA_max = ptRNA_list_slow[np.argmax(elongt_list_slow)]\n",
    "tRNA_distributions['GFP_gaslow'] = [elongt_slowest,dict(zip(tRNA_tags,ptRNA_max)),dict(zip(codon_tags,pCodon_GFP))]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.00606658, 0.        , 0.00228968, 0.        , 0.        ,\n",
       "       0.        , 0.00326699, 0.0043185 , 0.01338964, 0.        ,\n",
       "       0.00289458, 0.00679118, 0.00462071, 0.        , 0.00381738,\n",
       "       0.00304735, 0.        , 0.        , 0.00259271, 0.        ,\n",
       "       0.        , 0.00221045, 0.01085204, 0.01972292, 0.62466653,\n",
       "       0.        , 0.        , 0.        , 0.0024463 , 0.        ,\n",
       "       0.        , 0.        , 0.00356713, 0.        , 0.        ,\n",
       "       0.        , 0.27650342, 0.00404843, 0.        , 0.00288748])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ptRNA_list_slow[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
